src/autoschedulers/anderson2021/DefaultCostModel.cpp

```diff
@@ -0,0 +1,401 @@
+        if (std::isnan(dst(i))) {
+            any_nans = true;
+            aslog(0) << "Prediction " << i << " is NaN. True runtime is " << true_runtimes(i) << "\n";
+            aslog(0) << "Checking pipeline features for NaNs...\n";
+            pipeline_feat_queue.for_each_value([&](float f) { if (std::isnan(f)) abort(); });
+            aslog(0) << "None found\n";
+            aslog(0) << "Checking schedule features for NaNs...\n";
+            schedule_feat_queue.for_each_value([&](float f) { if (std::isnan(f)) abort(); });
+            aslog(0) << "None found\n";
+            aslog(0) << "Checking network weights for NaNs...\n";
+            weights.for_each_buffer([&](const Runtime::Buffer<float> &buf) {
+                buf.for_each_value([&](float f) { if (std::isnan(f)) abort(); });
+            });
+            aslog(0) << "None found\n";
+        }
+        internal_assert(true_runtimes(i) > 0);
+    }
+    if (any_nans) abort();
+
+    // Update weights locally
+    auto update_weight = [](const Runtime::Buffer<float> &src, Runtime::Buffer<float> &dst) {
+        dst.copy_from(src.sliced(src.dimensions() - 1, 0));
+    };
+    update_weight(head1_filter_update, weights.head1_filter);
+    update_weight(head1_bias_update, weights.head1_bias);
+    update_weight(head2_filter_update, weights.head2_filter);
+    update_weight(head2_bias_update, weights.head2_bias);
+    update_weight(conv1_filter_update, weights.conv1_filter);
+    update_weight(conv1_bias_update, weights.conv1_bias);
+
+    internal_assert(cursor != 0);
+
+    return loss();
+}
+
+void DefaultCostModel::evaluate_costs() {
+    if (cursor == 0 || !schedule_feat_queue.data()) return;
+
+    internal_assert(pipeline_feat_queue.data());
+    internal_assert(schedule_feat_queue.data());
+
+    Runtime::Buffer<float> dst = costs.cropped(0, 0, cursor);
+    Runtime::Buffer<float> dst_costs_per_stage = costs_per_stage.cropped({{0, cursor}, {0, num_stages}});
+
+    auto loss = Runtime::Buffer<float>::make_scalar();
+
+    int result = cost_model(num_stages,
+                            cursor,
+                            num_cores,
+                            batch_id++,
+                            pipeline_feat_queue,
+                            schedule_feat_queue,
+                            weights.head1_filter, weights.head1_bias,
+                            weights.head2_filter, weights.head2_bias,
+                            weights.conv1_filter, weights.conv1_bias,
+                            0.0f, 0, 0, nullptr,
+                            dst, dst_costs_per_stage, loss);
+    (void)result;
+    internal_assert(result == 0);
+
+    for (int i = 0; i < cursor; i++) {
+        internal_assert(cost_ptrs(i));
+        *(cost_ptrs(i)) = dst(i);
+        for (int s = 0; s < num_stages; ++s) {
+            (*cost_per_stage_ptrs[i])[s] = dst_costs_per_stage(i, s);
+        }
+    }
+
+    cursor = 0;
+    cost_per_stage_ptrs.clear();
```

Comment:
I might have been wrong, but it seems that the buildbots are not running the GPU autoscheduler related tests?
For example, I pulled this branch and tried running `anderson2021_test_apps_autoscheduler` (i.e., `test/autoschedulers/anderson2021/test.cpp`) but it instantly fails, complaining that access to `cost_per_stage_ptrs` is out of bound. I commented this line (`cost_per_stage_ptrs.clear();`) and it seems to work again.

Reply:
> I might have been wrong, but it seems that the buildbots are not running the GPU autoscheduler related tests?

That's correct and as-intended, since this branch hasn't landed yet. The buildbots usually aren't configured to test major new features before they have landed. This PR is close, but has a number of nits that need fixing first.